apps/protocol-frontend/src/components/SignatureLayout.tsx

```diff
@@ -0,0 +1,278 @@
+      ) {
+        await createUser({
+          newUser: {
+            username: displayNameFromParams ?? '', // see if displayname from discord exists and create with this username if so
+            address: address,
+          },
+          showToast: false,
+        });
+      }
+    }
+    const res = await fetch(`${BACKEND_ADDR}/discord_nonce`, {
+      method: 'GET',
+      credentials: 'include',
+    });
+    const state = await res.text();
+
+    const params = new URLSearchParams();
+    params.append('client_id', DISCORD_CLIENT_ID);
+    params.append('redirect_uri', DISCORD_REDIRECT_URI);
+    params.append('response_type', 'code');
+    params.append('scope', 'identify guilds.join');
+    params.append('state', `${state}/${address}/signature`); // generate string to prevent crsf attack and include signature as source
+    params.append('prompt', 'consent');
+    window.location.href = `https://discord.com/oauth2/authorize?${params.toString()}`;
+  };
+
+  return (
+    <Flex
+      direction="column"
+      alignItems={{ base: 'center' }}
+      justifyContent={{ base: 'center' }}
+      overflowX="hidden"
+      margin="0 auto"
+      paddingX={{ base: 8, lg: 0 }}
+      minHeight="100vh"
+      minWidth="100vw"
+      bgGradient="linear-gradient(49deg, rgba(223,31,151,1) 0%, rgba(81,0,228,1) 100%)"
+    >
+      <Flex
+        direction="column"
+        alignItems={{ base: 'center' }}
+        justifyContent={{ base: 'center' }}
+        paddingY={{ base: '8', lg: '0' }}
+        flex="1"
+        minHeight={['100vh', '100vh', '0', '0']}
+        gap={{ base: 4, lg: 16 }}
+      >
+        <VisuallyHidden>
+          <Heading as="h1">Govrn</Heading>
+        </VisuallyHidden>
+        <GovrnTextLogo />
+        <Flex
+          direction="column"
+          alignItems="center"
+          justifyContent="center"
+          color="white"
+          fontSize={{ base: 'lg', lg: '2xl' }}
+          fontWeight="400"
+          marginBottom={{ base: 40, lg: 0 }}
+          gap={1}
+          maxW={{ base: '60%', lg: '75%' }}
+        >
+          <Flex
+            direction="column"
+            alignItems="center"
+            justifyContent="center"
+            color="white"
+            fontSize={{ base: 'lg', lg: 'xl' }}
+            fontWeight="400"
+            maxW={{ base: '60%', lg: '75%' }}
+            textAlign="center"
+          >
+            {!isConnected && userDataByAddress === undefined ? ( // stage 1: connect wallet
+              <StageOne />
+            ) : null}
+            {isConnected && !isDiscordConnected(userData) ? ( // stage 2: discord oauth
+              userData ? (
+                <StageTwoExistingUser />
+              ) : (
+                <StageTwoNewUser />
+              )
+            ) : null}
+            {isConnected && userData && isDiscordConnected(userData) ? ( // stage 3: success and next steps
+              <StageThreeSuccess />
+            ) : null}
```

Comment:
We should move some of this outside of the JSX and possibly encapsulated in a component rather than having these nested ternaries.

Reply:
I agree. I want to revisit that - it's how I did it initially and I ran into some strange re renders so I moved to these.